### PR TITLE
initial work to add documentation of GPS WPF

### PR DESCRIPTION
**Pull request by jediofgever**
Related to https://github.com/ros-planning/navigation.ros.org/issues/127

### CODE REVIEW

**Comment by SteveMacenski** [comment]
You could skip the tutorial for the case of just the core things needed for the documentation of the PR. We could later add the tutorial so that this isn't blocking.

Or we could do it here too, it's up to you :smile:

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+==============
+
+0- Collect GPS Waypoints
+------------------------
+
+We can consider two cases here. GPS waypoint following is possible in both simulation and real. This tutorial was developed based on simulation but all the steps should also apply on a real system. 
+
+Although It might not be the most convinient way, a practical way to collect the GPS waypoints is to traverse the path and collect GPS waypoints along the way with a frequency.
+
+The action client node then can read this waypoints in yaml file and go through them. Assuming that you have ``sensor_msgs::msg::NavSatFix`` message type available under /gps/fix topic and ``sensor_msgs::msg::Imu``
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:39]: "collector" - also link to that

Comment (SteveMacenski) [reply tutorials/docs/navigation2_gps_navigation.rst:39]: "collect"

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+
+Although It might not be the most convinient way, a practical way to collect the GPS waypoints is to traverse the path and collect GPS waypoints along the way with a frequency.
+
+The action client node then can read this waypoints in yaml file and go through them. Assuming that you have ``sensor_msgs::msg::NavSatFix`` message type available under /gps/fix topic and ``sensor_msgs::msg::Imu``  
+message type available under /imu/data topic, you might drive the robot to points and use provided gps_waypoint_collectr node to colect GPS coordinates and robot's orientation at that point. 
+
+We are interested in getting latitude, longitude and altitude at this location. We also need to consider a desired orientation at this pose. A global orientation can be acquired from an earth referenced IMU. Please see more details on
+how the IMU should report the data at `Robot Localization <http://docs.ros.org/en/melodic/api/robot_localization/html/preparing_sensor_data.html#imu>`_. For the orientation, only yaw angle is our concern here.
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:43]:
```suggestion
how the IMU should report the data at `Robot Localization <http://docs.ros.org/en/melodic/api/robot_localization/html/preparing_sensor_data.html#imu>`_. 
```

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+
+The action client node then can read this waypoints in yaml file and go through them. Assuming that you have ``sensor_msgs::msg::NavSatFix`` message type available under /gps/fix topic and ``sensor_msgs::msg::Imu``  
+message type available under /imu/data topic, you might drive the robot to points and use provided gps_waypoint_collectr node to colect GPS coordinates and robot's orientation at that point. 
+
+We are interested in getting latitude, longitude and altitude at this location. We also need to consider a desired orientation at this pose. A global orientation can be acquired from an earth referenced IMU. Please see more details on
+how the IMU should report the data at `Robot Localization <http://docs.ros.org/en/melodic/api/robot_localization/html/preparing_sensor_data.html#imu>`_. For the orientation, only yaw angle is our concern here. 
+
+We have provided a node in tutorial source code to collect synced imu and gps data at waypoints. The topics can be configured to your case. ;
+
+``gps_waypoint_collector.launh.py`` in ``nav2_gps_waypoint_follower_demo`` ;
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:47]: "launch"

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+      return LaunchDescription([
+          launch_ros.actions.Node(
+                  package='nav2_gps_waypoint_follower_demo', 
+                  executable='gps_waypoint_collector', 
+                  name='gps_waypoint_collector_node',
+                  output='screen',
+                  remappings=[('/gps', '/gps/fix'),
+                              ('/imu', '/imu/data')],  
+                  parameters=[{'frequency': 1},  # dont set to high otherwise you will end up with hundreds of waypoints!
+                              {'yaml_file_out': "/home/user_name/collected_points.yaml"}]  #change this according to your home dir
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:71]:
```suggestion
                              {'yaml_file_out': "/path/to/collected_points.yaml"}]   # change this to desired location
```

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+              )               
+  ])
+
+
+remap your topics accordingly.
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:76]: Delete this

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+                              ('/imu', '/imu/data')],  
+                  parameters=[{'frequency': 1},  # dont set to high otherwise you will end up with hundreds of waypoints!
+                              {'yaml_file_out': "/home/user_name/collected_points.yaml"}]  #change this according to your home dir
+              )               
+  ])
+
+
+remap your topics accordingly. 
+
+frequency parameter will determine density of collected points. Also do not forget to replace your user name in the path given to yaml file.
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:78]: code block params

Comment (SteveMacenski) [reply tutorials/docs/navigation2_gps_navigation.rst:78]: These also don't all need to be single paragraphs, please combine more of these sentences into structure.

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+
+You need to make sure that you have built packages in `navigation2_tutorials <https://github.com/ros-planning/navigation2_tutorials>`_ successfully. 
+After you drove the robot to a waypoint then do ;
+
+.. code-block:: bash
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:83]: None of these code blocks are rendering

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+  [gps_waypoint_collector-1] [INFO] [1609821278.276561015] [gps_waypoint_collector_node]: curr_gps_waypoint: [0.00000000, 0.00000003, 0.63917288, 0.7]
+  [gps_waypoint_collector-1] [INFO] [1609821279.276582547] [gps_waypoint_collector_node]: curr_gps_waypoint: [0.00000000, 0.00000003, 0.63917288, 0.75]
+  [gps_waypoint_collector-1] [INFO] [1609821280.276421593] [gps_waypoint_collector_node]: curr_gps_waypoint: [0.00000000, 0.00000003, 0.63917288, 0.755]
+
+here the values corresponds to ; 
+
+.. code-block:: bash
+  curr_gps_waypoint: [lat, long, alt, yaw(radians)]  
+
+This callback is called periodicly accrding to the frequency you set, if the robot moves the values will be updated. The points are published to a topic as well.
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:100]: "periodically" 

Please review this tutorial a few times and find the typos - there's a bunch of them

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+  height: 80
+  resolution: 0.2
+
+Adjusts the sizes of global and local costmaps according to your desired reachability.   
+
+
+3- Let The Robot Follow The GPS Waypoints
+-----------------------------------------
+
+Assuming that you have already collected your waypoints and they reside at `navigation2_tutorials/nav2_gps_waypoint_follower_demo/params/demo_gps_waypoints.yaml`.
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:373]:
```suggestion
For the purposes of this section, we assume that you have already collected your waypoints and they reside at ``navigation2_tutorials/nav2_gps_waypoint_follower_demo/params/demo_gps_waypoints.yaml``. 
```

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+
+Adjusts the sizes of global and local costmaps according to your desired reachability.   
+
+
+3- Let The Robot Follow The GPS Waypoints
+-----------------------------------------
+
+Assuming that you have already collected your waypoints and they reside at `navigation2_tutorials/nav2_gps_waypoint_follower_demo/params/demo_gps_waypoints.yaml`. 
+
+You can simply launch the GPS waypoint following with provided node.
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:375]: These all don't need to be independent paragraphs

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+-----------------------------------------
+
+Assuming that you have already collected your waypoints and they reside at `navigation2_tutorials/nav2_gps_waypoint_follower_demo/params/demo_gps_waypoints.yaml`. 
+
+You can simply launch the GPS waypoint following with provided node. 
+
+You need to make sure that you have built packages in `navigation2_tutorials <https://github.com/ros-planning/navigation2_tutorials>`_ successfully. 
+After sucessful build do;
+
+.. code-block:: bash
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:380]: Not rendering

File: tutorials/docs/navigation2_gps_navigation.rst [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+                                  output='screen',
+                                  parameters=[parameter_file],
+                                  )
+
+      return LaunchDescription([
+          params_declare,
+          driver_node,
+      ])
+
+If you save more than one different waypoint data files then just change the demo_gps_waypoints.yaml with your desired file.
```
Comment (SteveMacenski) [inline tutorials/docs/navigation2_gps_navigation.rst:431]: This section doesn't really end with a robot following waypoints or explaining any of the things they're launching. Alot more explanation required.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Good first run at the tutorial, but it needs work. 
- Ordering needs some work to get across the GPS navigation points first and very clearly before we discuss the specific implementation of it using a waypoint follower
- Typos and missing renderings of code blocks - if you have CI build, you can go to circle -> artifacts ->html/index.html to see your rendered version. You can also compile locally. 
- Generally more discussion required on topics

**Comment by jediofgever** [comment]
I will work on items you mentioned, but my work's priority has changed so I cannot look into it on my work time but in free additional time.

**Comment by SteveMacenski** [comment]
Superseded!

**Pull request closed without merging**